File: market_wizards/templates/_components/canvas/canvas.dart
import "dart:html";
import "dart:math";
import "../../_services/server.dart";

class Canvas {
  final DivElement _info;

  //final DivElement _container;
  final ImageElement _image;

  final CanvasElement _inspectCanvas;
  final CanvasElement _coverCanvas;

  final Server _server;

  final String _cls;

  CanvasRenderingContext2D _ictx;
  CanvasRenderingContext2D _cctx;

  bool _calculate = false;
  Map<String, dynamic> _calculateStart;
  Map<String, dynamic> _calculateEnd;

  num _calcStartAnchorX = 0;
  num _calcStartAnchorY = 0;

  bool _coverL = false;
  bool _coverR = false;

  bool _double = false;
  num _doubleAnchorX = 0;

  Canvas(String id, {String cls = ""})
      : _cls = cls,
        _server = new Server(),
        _info = querySelector("#${id}-canvas-chart-info"),
        //_container = querySelector("#${id}-canvas-chart-container"),
        _image = querySelector("#${id}-canvas-chart-image"),
        _inspectCanvas = querySelector("#${id}-canvas-chart-inspect"),
        _coverCanvas = querySelector("#${id}-canvas-chart-cover") {
    _ictx = _inspectCanvas.getContext("2d");
    _cctx = _coverCanvas.getContext("2d");

    _attachListener();
  }

  String get _coverColor => "rgba(0, 0, 0, 0.7)";
  String get _inspectColor => "rgba(255, 255, 255, 0.8)";
  String get _anchorColor => "rgba(255, 255, 255, 0.5)";

  void initCanvasSize() {
    _inspectCanvas.width = _image.clientWidth.floor() - 1;
    _inspectCanvas.height = _image.clientHeight.floor() - 1;

    _coverCanvas.width = _image.clientWidth.floor() - 1;
    _coverCanvas.height = _image.clientHeight.floor() - 1;
  }

  void _attachListener() {
    window.onResize.listen((Event event) {
      initCanvasSize();
    });

    _image.onLoad.listen((Event event) {
      //_container.classes.remove("${_cls}-canvas-chart-hidden");
      initCanvasSize();

      _server.done();

      _server.timeRequest();

      if (!_info.classes.contains("${_cls}-canvas-chart-info-hidden")) {
        _info.classes.add("${_cls}-canvas-chart-info-hidden");
      }
    });

    _server.$chartUrl.listen((url) {
      //_container.classes.add("${_cls}-canvas-chart-hidden");
      _image.src = url;
    });

    _server.getChart();

    _server.$chartInspect.listen((info) {
      var str = "time: ${info['time']}\n"
          "price: ${info['price']}\n"
          "open: ${info['open']}\n"
          "high: ${info['high']}\n"
          "low: ${info['low']}\n"
          "close: ${info['close']}\n"
          "volume: ${info.putIfAbsent('volume', () => 0)}\n"
          "interest: ${info.putIfAbsent('open interest', () => 0)}\n";

      if (!_calculate) {
        _calculateStart = info;
      } else {
        _calculateEnd = info;

        var d = DateTime.parse(_calculateEnd["time"])
            .difference(DateTime.parse(_calculateStart["time"]));

        var p = num.parse(_calculateEnd["price"].replaceAll(",", "")) -
            num.parse(_calculateStart["price"].replaceAll(",", ""));

        var pp = (num.parse(_calculateEnd["price"].replaceAll(",", "")) -
                num.parse(_calculateStart["price"].replaceAll(",", ""))) /
            num.parse(_calculateStart["price"].replaceAll(",", ""));

        pp *= 100.0;

        var sd = d.inDays.toString().replaceAllMapped(
            new RegExp(r"(\d{1,3})(?=(\d{3})+)"), (Match m) => "${m[1]},");

        var re = new RegExp(r"(\d{1,3})(?=(\d{3})+\.)");

        var sp =
            p.toStringAsFixed(2).replaceAllMapped(re, (Match m) => "${m[1]},");

        var spp =
            pp.toStringAsFixed(2).replaceAllMapped(re, (Match m) => "${m[1]},");

        str = "${str}\nperiod: ${sd}days\ndiff(\$): ${sp}\ndiff(%): ${spp}";
      }

      _info.innerHtml = str;
    });

    document.body.onMouseMove.listen((MouseEvent event) {
      _inspectInfo(event);

      if (_double) {
        _doubleCover(event);
      } else if (_coverL) {
        _singleCoverL(event);
      } else if (_coverR) {
        _singleCoverR(event);
      } else {}
      _inspect(event);
      if (_calculate) {
        _calcAnchor(event);
      }
    });

    document.body.onMouseDown.listen((MouseEvent event) {
      _ictx.clearRect(0, 0, _inspectCanvas.width, _inspectCanvas.height);
      _cctx.clearRect(0, 0, _inspectCanvas.width, _inspectCanvas.height);

      if (!_info.classes.contains("${_cls}-canvas-chart-info-hidden")) {
        _info.classes.add("${_cls}-canvas-chart-info-hidden");
      }

      _calculate = true;
      _calcStartAnchorX = _eventXOffset(event);
      _calcStartAnchorY = _eventYOffset(event);

      if (event.ctrlKey) {
        _singleCoverL(event);
      }

      if (event.shiftKey) {
        _singleCoverR(event);
      }

      if (event.altKey) {
        _doubleCover(event);
      }
    });

    _coverCanvas.onMouseUp.listen((MouseEvent event) {
      _calculate = false;

      _coverL = false;
      _coverR = false;

      _double = false;
      _doubleAnchorX = 0;
    });
  }

  num _eventXOffset(MouseEvent event) {
    return event.client.x - _image.offsetLeft;
  }

  num _eventYOffset(MouseEvent event) {
    return event.client.y - _image.offsetTop;
  }

  void _singleCoverR(MouseEvent event) {
    _coverR = true;

    _cctx.clearRect(0, 0, _coverCanvas.width, _coverCanvas.height);
    _cctx.fillStyle = _coverColor;

    _cctx.fillRect(_eventXOffset(event), 0,
        _coverCanvas.width - _eventXOffset(event), _coverCanvas.height);
  }

  void _singleCoverL(MouseEvent event) {
    _coverL = true;

    _cctx.clearRect(0, 0, _coverCanvas.width, _coverCanvas.height);
    _cctx.fillStyle = _coverColor;

    _cctx.fillRect(0, 0, _eventXOffset(event), _coverCanvas.height);
  }

  void _doubleCover(MouseEvent event) {
    _cctx.clearRect(0, 0, _coverCanvas.width, _coverCanvas.height);
    _cctx.fillStyle = _coverColor;

    if (!_double) {
      _doubleAnchorX = _eventXOffset(event);
    }

    _double = true;

    if (_eventXOffset(event) >= _doubleAnchorX) {
      _cctx.fillRect(0, 0, _doubleAnchorX, _coverCanvas.height);

      _cctx.fillRect(_eventXOffset(event), 0,
          _coverCanvas.width - _eventXOffset(event), _coverCanvas.height);
    } else {
      _cctx.fillRect(0, 0, _eventXOffset(event), _coverCanvas.height);

      _cctx.fillRect(_doubleAnchorX, 0, _coverCanvas.width - _doubleAnchorX,
          _coverCanvas.height);
    }
  }

  void _inspectInfo(MouseEvent event) {
    if (_info.classes.contains("${_cls}-canvas-chart-info-hidden")) {
      _info.classes.remove("${_cls}-canvas-chart-info-hidden");
    }

    var x = max(min(_eventXOffset(event) / _inspectCanvas.width, 1), 0);
    var y = max(
        min(
            (_inspectCanvas.height - _eventYOffset(event)) /
                _inspectCanvas.height,
            1),
        0);

    _server.inspectRequest(x, y);

    var offset = 20;

    if (_eventXOffset(event) > _inspectCanvas.width / 2) {
      _info.style.left = "${event.client.x - _info.clientWidth - offset}px";
    } else {
      _info.style.left = "${event.client.x + offset}px";
    }

    if (_eventYOffset(event) > _inspectCanvas.height / 2) {
      _info.style.top = "${event.client.y - _info.offsetHeight - offset}px";
    } else {
      _info.style.top = "${event.client.y + offset}px";
    }
  }

  void _inspect(MouseEvent event) {
    _ictx.clearRect(0, 0, _inspectCanvas.width, _inspectCanvas.height);
    _ictx.strokeStyle = _inspectColor;

    _ictx.beginPath();
    _ictx.moveTo(_eventXOffset(event), 0);

    _ictx.lineTo(_eventXOffset(event), _inspectCanvas.height);

    _ictx.stroke();
    _ictx.closePath();

    _ictx.beginPath();

    _ictx.moveTo(0, _eventYOffset(event));

    _ictx.lineTo(_inspectCanvas.width, _eventYOffset(event));

    _ictx.stroke();
    _ictx.closePath();
  }

  void _calcAnchor(MouseEvent event) {
    _ictx.strokeStyle = _anchorColor;

    _ictx.beginPath();
    _ictx.moveTo(_calcStartAnchorX, 0);

    _ictx.lineTo(_calcStartAnchorX, _inspectCanvas.height);

    _ictx.stroke();
    _ictx.closePath();

    _ictx.beginPath();

    _ictx.moveTo(0, _calcStartAnchorY);

    _ictx.lineTo(_inspectCanvas.width, _calcStartAnchorY);

    _ictx.stroke();
    _ictx.closePath();
  }
}

File: market_wizards/react/src/components/canvas/canvas.js
import React, { useEffect, useRef, useState, useContext } from "react";
import styles from "./canvas.module.scss";
import { ChartContext } from "../../context/chart";

function Canvas() {
  const { addImageCallback, inspectRequest } = useContext(ChartContext);

  const coverColor = "rgba(0, 0, 0, 0.8)";
  const inspectColor = "rgba(255, 255, 255, 0.8)";
  const anchorColor = "rgba(255, 255, 255, 0.5)";

  const infoRef = useRef(null);
  const inspectRef = useRef(null);
  const coverRef = useRef(null);
  const imageRef = useRef(null);

  const params = useRef({
    left: false,
    right: false,
    both: false,
    calc: false,
    calcX: 0,
    calcY: 0
  });

  const [state, setState] = useState({
    moving: false
  });

  function eventXOffset(e) {
    return e.clientX - imageRef.current.offsetLeft;
  }

  function eventYOffset(e) {
    return e.clientY - imageRef.current.offsetTop;
  }

  function initCanvasSize() {
    inspectRef.current.width = Math.floor(imageRef.current.clientWidth) - 1;
    inspectRef.current.height = Math.floor(imageRef.current.clientHeight) - 1;

    coverRef.current.width = Math.floor(imageRef.current.clientWidth) - 1;
    coverRef.current.height = Math.floor(imageRef.current.clientHeight) - 1;
  }

  function singleCoverR(e) {
    const cctx = coverRef.current.getContext("2d");

    cctx.clearRect(0, 0, coverRef.current.width, coverRef.current.height);
    cctx.fillStyle = coverColor;

    cctx.fillRect(
      eventXOffset(e),
      0,
      coverRef.current.width - eventXOffset(e),
      coverRef.current.height
    );
  }

  function singleCoverL(e) {
    const cctx = coverRef.current.getContext("2d");

    cctx.clearRect(0, 0, coverRef.current.width, coverRef.current.height);
    cctx.fillStyle = coverColor;

    cctx.fillRect(0, 0, eventXOffset(e), coverRef.current.height);
  }

  function doubleCover(e) {
    const cctx = coverRef.current.getContext("2d");

    cctx.clearRect(0, 0, coverRef.current.width, coverRef.current.height);
    cctx.fillStyle = coverColor;

    if (eventXOffset(e) >= params.current.calcX) {
      cctx.fillRect(0, 0, params.current.calcX, coverRef.current.height);

      cctx.fillRect(
        eventXOffset(e),
        0,
        coverRef.current.width - eventXOffset(e),
        coverRef.current.height
      );
    } else {
      cctx.fillRect(0, 0, eventXOffset(e), coverRef.current.height);

      cctx.fillRect(
        params.current.calcX,
        0,
        coverRef.current.width - params.current.calcX,
        coverRef.current.height
      );
    }
  }

  function inspectInfo(e) {
    const x = Math.max(
      Math.min(eventXOffset(e) / inspectRef.current.width, 1),
      0
    );
    const y = Math.max(
      Math.min(
        (inspectRef.current.height - eventYOffset(e)) /
          inspectRef.current.height,
        1
      ),
      0
    );

    if (params.current.calc) {
      const ax = Math.max(
        Math.min(params.current.calcX / inspectRef.current.width, 1),
        0
      );

      const ay = Math.max(
        Math.min(
          (inspectRef.current.height - params.current.calcY) /
            inspectRef.current.height,
          1
        ),
        0
      );

      inspectRequest(
        data => {
          infoRef.current.innerHTML = data;
        },
        x,
        y,
        ax,
        ay
      );
    } else {
      inspectRequest(
        data => {
          infoRef.current.innerHTML = data;
        },
        x,
        y
      );
    }

    const offset = 20;

    let l;
    if (eventXOffset(e) > inspectRef.current.width / 2) {
      l = `${e.clientX - infoRef.current.clientWidth - offset}px`;
    } else {
      l = `${e.clientX + offset}px`;
    }

    infoRef.current.style.left = l;

    let t;
    if (eventYOffset(e) > inspectRef.current.height / 2) {
      t = `${e.clientY - infoRef.current.offsetHeight - offset}px`;
    } else {
      t = `${e.clientY + offset}px`;
    }

    infoRef.current.style.top = t;
  }

  function inspect(e) {
    const ictx = inspectRef.current.getContext("2d");

    ictx.clearRect(0, 0, inspectRef.current.width, inspectRef.current.height);
    ictx.strokeStyle = inspectColor;

    ictx.beginPath();
    ictx.moveTo(eventXOffset(e), 0);

    ictx.lineTo(eventXOffset(e), inspectRef.current.height);

    ictx.stroke();
    ictx.closePath();

    ictx.beginPath();

    ictx.moveTo(0, eventYOffset(e));

    ictx.lineTo(inspectRef.current.width, eventYOffset(e));

    ictx.stroke();
    ictx.closePath();
  }

  function calcAnchor() {
    const ictx = inspectRef.current.getContext("2d");

    ictx.strokeStyle = anchorColor;

    ictx.beginPath();
    ictx.moveTo(params.current.calcX, 0);

    ictx.lineTo(params.current.calcX, inspectRef.current.height);

    ictx.stroke();
    ictx.closePath();

    ictx.beginPath();

    ictx.moveTo(0, params.current.calcY);

    ictx.lineTo(inspectRef.current.width, params.current.calcY);

    ictx.stroke();
    ictx.closePath();
  }

  function handlerUp() {
    params.current.left = false;
    params.current.right = false;
    params.current.both = false;
    params.current.calc = false;
  }

  function handlerMove(e) {
    if (!state.moving) {
      setState({ ...state, moving: true });
    }

    inspectInfo(e);

    if (params.current.both) {
      doubleCover(e);
    } else if (params.current.left) {
      singleCoverL(e);
    } else if (params.current.right) {
      singleCoverR(e);
    }

    inspect(e);

    if (params.current.calc) {
      calcAnchor(e);
    }
  }

  function handlerDown(e) {
    const ictx = inspectRef.current.getContext("2d");
    const cctx = coverRef.current.getContext("2d");

    ictx.clearRect(0, 0, inspectRef.current.width, inspectRef.current.height);
    cctx.clearRect(0, 0, inspectRef.current.width, inspectRef.current.height);

    params.current.left = false;
    params.current.right = false;
    params.current.both = false;

    if (e.ctrlKey) {
      params.current.left = true;
    } else if (e.shiftKey) {
      params.current.right = true;
    } else if (e.altKey) {
      params.current.both = true;
    }

    params.current.calc = true;
    params.current.calcX = eventXOffset(e);
    params.current.calcY = eventYOffset(e);

    setState({
      ...state,
      moving: false
    });
  }

  useEffect(() => {

    addImageCallback("CANVAS", img => {
      imageRef.current.src = img
    })

    const imgLoaded = () => {
      initCanvasSize();
    };

    const irf = imageRef.current;

    irf.addEventListener("load", imgLoaded);

    window.addEventListener("resize", () => {
      initCanvasSize();
    });

    window.addEventListener("mouseup", handlerUp);
    window.addEventListener("mousemove", handlerMove);
    window.addEventListener("mousedown", handlerDown);

    return () => {
      irf.addEventListener("load", imgLoaded);
      window.removeEventListener("mouseup", handlerUp);
      window.removeEventListener("mousemove", handlerMove);
      window.removeEventListener("mousedown", handlerDown);
    };
  });

  console.log("canvas");

  return (
    <>
      <div
        ref={infoRef}
        className={
          state.moving
            ? styles.chartInfo
            : `${styles.chartInfo} ${styles.chartInfoHidden}`
        }
      >
        Info
      </div>
      <div className={styles.mainContainer}>
        <div className={styles.chartContainer}>
          <canvas ref={inspectRef} className={styles.chartCover}></canvas>
          <canvas ref={coverRef} className={styles.chartCover}></canvas>

          <img ref={imageRef} className={styles.chartImage} src="" />
        </div>
      </div>
    </>
  );
}

function shouldUpdate(prev, next) {
  return prev.src !== next.src;
}

export default Canvas;
